Separating quantization test from distributed_test

marjanfariborz opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63154
* #63059
* __->__ #63058

Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Commit:
Separating quantization test from distributed_test

Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]

Commit:
Update on "Separating quantization test from distributed_test"


Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]

wanchaol commented on test/distributed/algorithms/quantization/quantization_test.py:
```diff
@@ -0,0 +1,108 @@
+    if device_id is None:
+        return torch.empty(size, size, size, dtype=dtype).fill_(value)
+    else:
+        return torch.empty(size, size, size, dtype=dtype).fill_(value).cuda(device_id)
```
nit: space between function def and class def?

wanchaol commented on test/distributed/algorithms/quantization/quantization_test.py:
```diff
@@ -0,0 +1,108 @@
+        return torch.empty(size, size, size, dtype=dtype).fill_(value)
+    else:
+        return torch.empty(size, size, size, dtype=dtype).fill_(value).cuda(device_id)
+class DistQuantizationTests(TestDistBackend, DistributedTest._DistTestBase):
```
what's the reason of importing these two base class for the collective tests? I think we can just base on `MultiProcessTestCase` similar to https://github.com/pytorch/pytorch/blob/master/test/distributed/algorithms/ddp_comm_hooks/test_ddp_hooks.py#L57 and https://github.com/pytorch/pytorch/blob/master/test/distributed/_sharded_tensor/test_sharded_tensor.py#L118 ?

wanchaol replied:
I see, but if you are trying to test different backends, shouldn't we skip some of the tests in certain backend? i.e. `all_to_all`  or `all_to_all_single` is not available in the gloo backend, we should skip it? we should ensure the collective only runs on the supported backend, we can ensure it like this https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/distributed/distributed_test.py#L860

marjanfariborz replied:
I thought `@requires_nccl()` would automatically do that. Can I use `sandcastle_skip_if` in this case?

wanchaol replied:
`requires_nccl` just ensure if `nccl` is not built together, the test get skipped, but it doesn't ensure we skip other backends properly, i.e. it might not skip the gloo backend for `all_to_all`. Yeah I think sandcastle_skip_if should be enough

marjanfariborz replied:
I see, okay then I will also add `sandcastle_skip_if`.

Commit:
Update on "Separating quantization test from distributed_test"


Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]

wanchaol reviewed (approved):
looks good, please ensure CI passing before landing

Commit:
Update on "Separating quantization test from distributed_test"


Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]

Commit:
Update on "Separating quantization test from distributed_test"


Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]

Commit:
Update on "Separating quantization test from distributed_test"


Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]

Commit:
Update on "Separating quantization test from distributed_test"


Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]

rohan-varma commented on test/distributed/algorithms/quantization/test_quantization.py:
```diff
@@ -0,0 +1,111 @@
+class DistQuantizationTests(TestDistBackend, DistributedTest._DistTestBase):
+    def setUp(self):
+        super().setUp()
+        self._fork_processes()
```
nit: I believe we are trying to move towards spawn over fork (various testing and deadlock issues with fork):
```
self._spawn_processes()
```

should be all you need.

marjanfariborz replied:
Done! thanks 😊

marjanfariborz replied:
When I used `self._spawn_processes()` all of my local tests passed  but the CI tests started complaining. But when I switched back to fork they passed. If that's okay I will land this code with fork and then use a separate commit for spawn.

rohan-varma reviewed (approved):
Thanks for enhancing the test suite!

Commit:
Update on "Separating quantization test from distributed_test"


Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]

Commit:
Update on "Separating quantization test from distributed_test"


Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]

Commit:
Update on "Separating quantization test from distributed_test"


Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]

Commit:
Update on "Separating quantization test from distributed_test"


Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]

Commit:
Update on "Separating quantization test from distributed_test"


Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]

Commit:
Update on "Separating quantization test from distributed_test"


Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]

Commit:
Update on "Separating quantization test from distributed_test"


Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]

Commit:
Update on "Separating quantization test from distributed_test"


Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]

Commit:
Update on "Separating quantization test from distributed_test"


Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]

wanchaol commented on test/distributed/algorithms/quantization/test_quantization.py:
```diff
@@ -0,0 +1,158 @@
+        store = dist.FileStore(self.file_name, self.world_size)
+        dist.init_process_group(store=store, rank=self.rank, world_size=self.world_size, backend='gloo')
+        device = torch.device(f"cuda:{self.rank}")
+        group_id = dist.new_group(range(self.world_size))
```
why we need to create a new_group here? we should use the global group instead, just like `_init_global_test` in `distributed_test.py` did?

wanchaol commented on test/distributed/algorithms/quantization/test_quantization.py:
```diff
@@ -0,0 +1,158 @@
+        store = dist.FileStore(self.file_name, self.world_size)
+        dist.init_process_group(store=store, rank=self.rank, world_size=self.world_size, backend='nccl')
+        device = torch.device(f"cuda:{self.rank}")
+        group_id = dist.new_group(range(self.world_size))
```
same

Commit:
Update on "Separating quantization test from distributed_test"


Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]

Commit:
Update on "Separating quantization test from distributed_test"


Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]

Commit:
Update on "Separating quantization test from distributed_test"


Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]

Commit:
Update on "Separating quantization test from distributed_test"


Dedicating separate tests for different quantization methods. Currently supporting FP16 method.

Differential Revision: [D30142580](https://our.internmc.facebook.com/intern/diff/D30142580/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30142580/)!

[ghstack-poisoned]
